pkg/code-generator/cmd/main.go

```diff
@@ -165,61 +156,96 @@ func collectProjectsFromRoot(root string, skipDirs []string) ([]model.ProjectCon
 	return projects, nil
 }
 
-func collectProtosFromRoot(root, tmpFile string, customImports, skipDirs []string, wantCompile func(string) bool) ([]*descriptor.FileDescriptorProto, error) {
-	var (
-		descriptors []*descriptor.FileDescriptorProto
-	)
+func addDescriptorsForFile(descriptors *[]*descriptor.FileDescriptorProto, root, protoFile string, customImports []string, wantCompile func(string) bool) error {
+	log.Printf("processing proto file input %v", protoFile)
+	imports, err := importsForProtoFile(root, protoFile, customImports)
+	if err != nil {
+		return errors.Wrapf(err, "reading imports for proto file")
+	}
+	imports = stringutils.Unique(imports)
+
+	// don't generate protos for non-project files
+	compile := wantCompile(protoFile)
+
+	// use a temp file to store the output from protoc, then parse it right back in
+	// this is how we "wrap" protoc
+	tmpFile, err := ioutil.TempFile("", "solo-kit-gen-")
+	if err != nil {
+		return err
+	}
+	if err := tmpFile.Close(); err != nil {
+		return err
+	}
+	defer os.Remove(tmpFile.Name())
+
+	if err := writeDescriptors(protoFile, tmpFile.Name(), imports, compile); err != nil {
+		return errors.Wrapf(err, "writing descriptors")
+	}
+	desc, err := readDescriptors(tmpFile.Name())
+	if err != nil {
+		return errors.Wrapf(err, "reading descriptors")
+	}
+
+addFiles:
+	for _, f := range desc.File {
+		// don't add the same proto twice, this avoids the issue where a dependency is imported multiple times
+		// with different import paths
+		for _, existing := range *descriptors {
+			if existing.GetName() == f.GetName() {
+				continue
+			}
+			existingCopy := proto.Clone(existing).(*descriptor.FileDescriptorProto)
+			existingCopy.Name = f.Name
+			if proto.Equal(existingCopy, f) {
+				continue addFiles
+			}
+		}
+		*descriptors = append(*descriptors, f)
```

Comment:
this is not thread safe as this method is called in parallel.
a simple (though ugly) fix would be to provide this function with a callback:
```
addDescriptor func(*descriptor.FileDescriptorProto)
```
and that function will be defined in `collectDescriptorsFromRoot` and will use a mutex:
```
       var descriptors []*descriptor.FileDescriptorProto 
       var mutex sync.Mutex
       addDescriptor := func(f *descriptor.FileDescriptorProto) {
          mutex.Lock()
	 descriptors = append(descriptors, f)
          mutex.Unlock()
       }
```